File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adjacency = new ArrayList<>();
        Queue<Integer> queue = new ArrayDeque<>();
        int[] result = new int[numCourses];
        int index = 0;
        for (int i = 0; i < numCourses; i++) {
            adjacency.add(new ArrayList<>());
        }
        int[] inDegree = new int[numCourses];
        for (int[] edge : prerequisites) {
            adjacency.get(edge[1]).add(edge[0]);
            inDegree[edge[0]]++;
        }

        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }
        while (!queue.isEmpty()) {
            int node = queue.poll();
            result[index++] = node;
            for (int neighbour : adjacency.get(node)) {
                inDegree[neighbour]--;
                if (inDegree[neighbour] == 0) {
                    queue.offer(neighbour);
                }
            }
        }

        return index == numCourses ? result : new int[0];
    }

}

File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int n, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(n);
        vector<int> vis(n);
        vector<int> ans;
        vector<int> in(n);

        for(int i=0; i<prerequisites.size(); i++){
            graph[prerequisites[i][1]].push_back(prerequisites[i][0]);
            in[prerequisites[i][0]]++;
        }
        
        queue<int> q;

        for(int i=0; i<n; i++){
            if(in[i]==0)q.push(i);
        }
        while(!q.empty()){
            int e = q.front();
            vis[e]=1;
            ans.push_back(e);
            q.pop();
            for(int x=0; x<graph[e].size(); x++){
                if(vis[graph[e][x]]==0){ 
                    in[graph[e][x]]--;
                    if(in[graph[e][x]]==0)q.push(graph[e][x]);
                }
            }
        }
        return ans.size() == n ? ans : vector<int>();
    }
};